flows/channel.go

```diff
@@ -54,66 +97,75 @@ func (c *channel) String() string {
 
 var _ utils.VariableResolver = (*channel)(nil)
 
-//------------------------------------------------------------------------------------------
-// JSON Encoding / Decoding
-//------------------------------------------------------------------------------------------
+// ChannelSet defines the unordered set of all channels for a session
+type ChannelSet struct {
+	channels       []Channel
+	channelsByUUID map[ChannelUUID]Channel
+}
 
-type channelEnvelope struct {
-	UUID        ChannelUUID `json:"uuid"`
-	Name        string      `json:"name"`
-	Address     string      `json:"address"`
-	ChannelType ChannelType `json:"type"`
+func NewChannelSet(channels []Channel) *ChannelSet {
+	s := &ChannelSet{channels: channels, channelsByUUID: make(map[ChannelUUID]Channel, len(channels))}
+	for _, channel := range s.channels {
+		s.channelsByUUID[channel.UUID()] = channel
+	}
+	return s
 }
 
-// ReadChannels decodes channels from the passed in JSON
-func ReadChannels(data []json.RawMessage) ([]Channel, error) {
-	channels := make([]Channel, len(data))
-	var err error
-	for c := range data {
-		channels[c], err = ReadChannel(data[c])
-		if err != nil {
-			return nil, err
+func (s *ChannelSet) WithRole(role ChannelRole) []Channel {
```

Comment:
to me `with_` methods generally should return the same type as the base object, so this is a bit weird to me. Might be able to just simplify all this by sticking the above method used for finding the right sending channel on ChannelSet entirely.

Reply:
Moved to `ChannelSet.GetForURN`